sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/EventProcessor{TPartition}.cs

```diff
@@ -1066,12 +1075,144 @@ async Task performProcessing()
-            await Task.Delay(Timeout.Infinite, cancellationToken).ConfigureAwait(false);
+            cancellationToken.ThrowIfCancellationRequested<TaskCanceledException>();
+
+            // Perform the tasks needed for the main processor loop, encompassing load balancing and management
+            // of processing for claimed partitions.
+            //
+            // Ensure that any errors observed, whether fatal or non-fatal, are surfaced to the exception handler in
+            // a fire-and-forget manner.  The processor does not assume responsibility for observing or surfacing exceptions
+            // that may occur in the handler, so the associated task is intentionally unobserved.
+
+            try
+            {
+                var connection = CreateConnection();
+                await using var connectionAwaiter = connection.ConfigureAwait(false);
+
+                var retryPolicy = Options.RetryOptions.ToRetryPolicy();
+                var cycleDuration = new Stopwatch();
+                var partitionIds = default(string[]);
+
+                while (!cancellationToken.IsCancellationRequested)
+                {
+                    cycleDuration.Restart();
+
+                    try
+                    {
+                        partitionIds = await connection.GetPartitionIdsAsync(retryPolicy, cancellationToken).ConfigureAwait(false);
+                    }
+                    catch (Exception ex) when (ex.IsNotType<TaskCanceledException>())
+                    {
+                        // Logging for exceptions with the service operation are responsibility of the connection.
+
+                        _ = InvokeOnProcessingErrorAsync(ex, null, Resources.OperationGetPartitionIds, CancellationToken.None);
+                        partitionIds = default;
+                    }
+
+                    // Perform a load balancing cycle only in the case that partitions were successfully read.
+
+                    if (partitionIds != default)
+                    {
+                        var claimedOwnership = default(EventProcessorPartitionOwnership);
+
+                        try
+                        {
+                            claimedOwnership = await LoadBalancer.RunLoadBalancingAsync(partitionIds, cancellationToken).ConfigureAwait(false);
+                        }
+                        catch (EventHubsException ex)
+                            when (Resources.OperationClaimOwnership.Equals(ex.GetFailureOperation(), StringComparison.InvariantCultureIgnoreCase))
+                        {
+                            // If a specific partition was associated with the failure, it is carried by a well-known data member of the exception.
+
+                            var partitionId = ex.GetFailureData<string>();
+
+                            var partition = (partitionId ?? string.Empty) switch
+                            {
+                                string id when (id == string.Empty) => null,
```

Comment:
For some reason, using `string.IsNullOrEmpty(partitionId)` here would not trigger this path when it was `null`.    I had to use a local capture here with an explicit `""` as a result.  I'd be interested in any insight that folks have on why that may have been the case.

Reply:
I'm a big fan of switch expressions/pattern matching.   While I acknowledge that it still looks weird to many, since it was only recently introduced in C# 8, but I expect that we'll see that getting more familiar going forward.  Many of the same concerns were raised around the use of pattern matching when it was first introduced, for example, and all the way back to generics.

Since we're fully embracing the C# 8 features within the codebase, including the use of switch expressions elsewhere, I'd prefer to keep it as is.  That said, I'm open if you feel strongly about it, though we'll probably want to extract out to a helper method in that case to avoid the `catch` block getting too chunky.  _(it's already borderline, in my opinion)_

Reply:
Don't feel super strongly.